src/db/couchPushEvents.ts

```diff
@@ -0,0 +1,328 @@
+  asObject({
+    eventId: asString, // Not the document id!
+    deviceId: asOptional(asString),
+    loginId: asOptional(asBase64),
+
+    // Event:
+    broadcastTxs: asOptional(asArray(asBroadcastTx)),
+    pushMessage: asOptional(asPushMessage),
+    recurring: asBoolean,
+    trigger: asPushTrigger,
+
+    // Status:
+    broadcastTxErrors: asOptional(asArray(asEither(asString, asNull))),
+    pushMessageEmits: asOptional(asNumber),
+    pushMessageFails: asOptional(asNumber),
+    state: asPushEventState,
+    triggered: asOptional(asDate)
+  })
+)
+const wasCouchPushEvent = uncleaner(asCouchPushEvent)
+
+/**
+ * Looks up events attached to devices.
+ */
+const deviceIdDesign = makeJsDesign('deviceId', ({ emit }) => ({
+  map: function (doc) {
+    if (doc.deviceId == null) return
+    if (doc.state === 'cancelled') return
+    if (doc.state === 'hidden') return
+    emit(doc.deviceId, null)
+  }
+}))
+
+/**
+ * Looks up events attached to logins.
+ */
+const loginIdDesign = makeJsDesign('loginId', ({ emit }) => ({
+  map: function (doc) {
+    if (doc.loginId == null) return
+    if (doc.state === 'cancelled') return
+    if (doc.state === 'hidden') return
+    emit(doc.loginId, null)
+  }
+}))
+
+/**
+ * Looks up active address-balance events.
+ */
+const addressBalanceDesign = makeJsDesign('address-balance', ({ emit }) => ({
+  map: function (doc) {
+    if (doc.trigger == null) return
+    if (doc.trigger.type !== 'address-balance') return
+    if (doc.state !== 'waiting') return
+    emit(doc._id, null)
+  }
+}))
+
+/**
+ * Looks up active price-related events.
+ */
+const priceDesign = makeJsDesign('price', ({ emit }) => ({
+  map: function (doc) {
+    if (doc.trigger == null) return
+    const type = doc.trigger.type
+    if (type !== 'price-change' && type !== 'price-level') return
+    if (doc.state !== 'waiting') return
+    emit(doc._id, null)
+  }
+}))
+
+/**
+ * Looks up active price-related events.
+ */
+const txConfirmDesign = makeJsDesign('tx-confirm', ({ emit }) => ({
+  map: function (doc) {
+    if (doc.trigger == null) return
+    if (doc.trigger.type !== 'tx-confirm') return
+    if (doc.state !== 'waiting') return
+    emit(doc._id, null)
+  }
+}))
+
+export const couchEventsSetup: DatabaseSetup = {
+  name: 'push-events',
+
+  documents: {
+    '_design/address-balance': addressBalanceDesign,
+    '_design/deviceId': deviceIdDesign,
+    '_design/loginId': loginIdDesign,
+    '_design/price': priceDesign,
+    '_design/tx-confirm': txConfirmDesign
+  }
+}
+
+export async function addEvent(
+  connection: ServerScope,
+  event: PushEvent,
+  created: Date
+): Promise<void> {
+  const db = connection.use(couchEventsSetup.name)
+  try {
+    await db.insert(
+      wasCouchPushEvent({
+        doc: event,
+        id: created.toISOString()
+      })
+    )
+  } catch (error) {
+    if (asMaybeConflictError(error) == null) throw error
+    await addEvent(connection, event, new Date(created.valueOf() + 1))
```

Comment:
We have been using this strategy in the other servers. It works quite well.

The basic idea is that at least one thread will always be making progress. If thread A, B, and C are all trying to add events at the same time, one will succeed and the other two will fail. That progress means we now we only have 2 events to add. So, the two remaining threads loop and and another one succeeds. Now we have only one event waiting, which will succeed on the third go-around. Even if more events arrive in the meantime, progress will continue happening.

This will only fail if we have more than 1000 events arrive per second. I am pretty sure CouchDB / CPU usage will die first.